File: cci05.03-reverse-bits-lcci/cpp/Solution.cpp
class Solution {
public:
    int reverseBits(int num) {
        if (num == 0) return 1;
        std::vector<int> nums(32);
        for (int i = 0; i < 32; ++i) {
            nums[i] = (num & 1);
            num >>= 1;
        }
        vector<int> leftOneCounts(32);
        int count = 0;
        for (int i = 0; i < 32; ++i) {
            leftOneCounts[i] = count;
            if (nums[i] == 1) {
                count++;
            } else {
                count = 0;
            }
        }
        vector<int> rightOneCounts(32);
        count = 0;
        for (int i = 31; i >= 0; --i) {
            rightOneCounts[i] = count;
            if (nums[i] == 1) {
                count++;
            } else {
                count = 0;
            }
        }
        int ret = 0;
        for (int i = 0; i < 32; ++i) {
            if (ret < leftOneCounts[i] + rightOneCounts[i] + 1) {
                ret = leftOneCounts[i] + rightOneCounts[i] + 1;
            }
        }
        return ret;
    }
};

File: cci05.03-reverse-bits-lcci/python/Solution.py
class Solution:
    def reverseBits(self, num: int) -> int:
        if num == 0:
            return 1
        nums = [None] * 32
        for i in range(32):
            nums[i] = num & 1
            num >>= 1
        leftOneCounts = [0] * 32
        count = 0
        for i in range(32):
            leftOneCounts[i] = count
            if nums[i] == 1:
                count += 1
            else:
                count = 0
        rightOneCounts = [0] * 32
        count = 0
        for i in range(31, -1,-1):
            rightOneCounts[i] = count
            if nums[i] == 1:
                count += 1
            else:
                count = 0
        ret = 0
        for i in range(32):
            if ret < leftOneCounts[i] + rightOneCounts[i] + 1:
                ret = leftOneCounts[i] + rightOneCounts[i] + 1
        return ret
